Add basic vault docs (#138)
timewave-computer/valence-protocol: docs/src/examples/crosschain_vaults.md

# Crosschain Vaults

> **Note:** _This example is still in design and includes new or experimental features of Valence Programs that may not be supported in current production releases._

Reviewer: ```suggestion
**Note:** _This example is still in the design phase and includes new or experimental features of Valence Programs that may not be supported in the current production release._
```

## Overview

> You can use Valence Programs to create crosschain vaults. Users interact with a vault on one chain while the hokens are held on another chain where yield is generated.

Reviewer: ```suggestion
You can use Valence Programs to create crosschain vaults. Users interact with a vault on one chain while the tokens are held on another chain where yield is generated.
```

Since gas is cheaper on Neutron than on Ethereum, computationally expensive operations, such as constraining the Strategist actions will be done on Neutron. Authorized messages will then be executed by each chain's Processor. Hyperlane is used to pass messages from the Authorization contract on Neutron to the Processor on Ethereum.

```mermaid
---
title: Program Control
---
graph LR
	Strategist
	subgraph Ethereum
		EP(Processor)
		EHM(Hyperlane Mailbox)
		EL(Ethereum Libraries)
		EVA(Valence Accounts)

	end
	subgraph Neutron
		A(Authorizations)
		NP(Processor)
		NHM(Hyperlane Mailbox)
		NL(Neutron Libraries)
		NVA(Valence Accounts)
	end

	Strategist --> A
	A --> NHM -- Relayer --> EHM --> EP --> EL --> EVA
	A --> NP --> NL--> NVA
```

### Libraries and Accounts needed

On Ethereum, we'll need Accounts for:
- **Deposit**: To hold user deposited tokens. Tokens from this pool can be then transported to Neutron.
- **Withdraw**: Told hold tokens received from Neutron. Tokens from this pool can then be 

On Neutron, we'll need Accounts for:

> - **Deposit**: To hold tokens bridged from Ethereum. Tokens from this pool can be used to enter into the position on Neutron.

Reviewer: ```suggestion
- **Position Depositor**: To hold tokens bridged from Ethereum. Tokens from this pool can be used to enter into the position on Neutron.
```

Author: Rejecting this suggestion. Position Depositor (verb) is a library. Deposit or Deposit Account is a pool of tokens that are available to Deposit.

> - **Position Holder**: Will hold the vouchers or shares associated with the position on Neutron

Reviewer: ```suggestion
- **Holding a position**: Will hold the vouchers or shares associated with the position on Neutron
```

> - **Position Holder**: Will hold the vouchers or shares associated with the position on Neutron

Reviewer: Holding

> - **Withdraw**: To hold the tokens that are withdrawn from the position. Tokens from this pool can be bridged back to Ethereum.

Reviewer: ```suggestion
- **Position Withdrawer**: To hold the tokens that are withdrawn from the position. Tokens from this pool can be bridged back to Ethereum.
```

Author: Similar to above.

  ```
  Pause {}
  ```
### Program subroutines

The program authorizes the Strategist to update the redemption rate and transport funds between various Accounts.

#### Allowing the Strategist to transport funds

```mermaid
---
title: From Ethereum Deposit Account to Neutron Position Account
---
graph LR
	subgraph Ethereum
		ED((Deposit))
		ET(Bridge Transfer)
	end
	subgraph Neutron
		NPH((Position Holder))
		NPD(Position Depositor)
		ND((Deposit))
	end

	ED --> ET --> ND --> NPD --> NPH
```

```mermaid
---
title: From Neutron Position Account to Ethereum Withdraw Account
---
graph RL
	subgraph Ethereum
		EW((Withdraw))
	end
	subgraph Neutron
		NPH((Position Holder))
		NW((Widthdraw))
		NT(Bridge Transfer)
		NPW(Position Withdrawer)
	end

	NPH --> NPW --> NW --> NT --> EW

```

```mermaid
---
title: Between Ethereum Deposit and Ethereum Withdraw Accounts
---
graph
	subgraph Ethereum
		ED((Deposit))
		EW((Withdraw))
		FDW(Forwarder)
	end
	ED --> FDW --> EW
```

## Design notes
This is a simplified design to demonstrate how a cross-chain vault may be implemented with Valence Programs. Production deployments will need to consider factors not covered here including:

> - Fees for gas, bridging, and for entering/exiting the position on Neutron. It is recommend that the vault impose withdraw fee and platform for users.

Author: Yes platform fee, i.e., a fee taken by the platform. It's taken not on exit or entry but annually on the total amount managed per user.
